### ISSUE TITLE
[FR] Log confusion matrix

### ISSUE THREAD

**Issue by robmarkcole** [microsoft/torchgeo#3376]
### Summary

To complement metrics. Use torchmetrics implementation? How to handle many classes?

**Comment by adamjstewart** [comment]
My suggestion was to plot a confusion matrix via tensorboard instead of logging all N x N values.

**Comment by DimitrisMantas** [comment]
I think perhaps implementing a plotting callback would be nice to do first and it would also help remove a bunch of boilerplate from the trainers...

**Comment by adamjstewart** [comment]
I actually love this idea, and I think it would also help a lot with @robmarkcole's other request: #2138. @DimitrisMantas do you want to try to take a stab at this? We would add a new `torchgeo/callbacks` directory similar to @hkristen's #3214.

**Comment by DimitrisMantas** [comment]
Sure, I can have a got at it. I'm just not sure what the scope of such a class should be. Supporting all possible loggers and plotable objects seems like too big a step right now.

Perhaps we could start with a metric plotter that accepts or assumes certain trainer fields, like \<stage\>\_confmat, and plots them on on TensorBoard at the right time. This way, the metric mixin can handle updates, and we won't have to change the on\_\<stage\>\_step return signatures, since if we were to plot predictions, like the segmentation trainer does, we would have to return whatever the callback would need to access.

**Comment by adamjstewart** [comment]
> I'm just not sure what the scope of such a class should be. Supporting all possible loggers and plotable objects seems like too big a step right now.

Maybe a base class for most shared code and a subclass for each logger? Just to be clear, I'm not asking you to implement a subclass for all loggers, just one or two that you care about (preferably TensorBoard to maintain easy backwards-compatibility).

> Perhaps we could start with a metric plotter that accepts or assumes certain trainer fields, like `<stage>_confmat`, and plots them on on TensorBoard at the right time. This way, the metric mixin can handle updates, and we won't have to change the `on_<stage>_step` return signatures, since if we were to plot predictions, like the segmentation trainer does, we would have to return whatever the callback would need to access.

Oof, I did not think about this. I don't want to do anything too custom. What do the Lightning docs suggest for callbacks? Should they have access to arbitrary instance attributes, or is it better to return all required values? For training, we must return loss, but for val (where we currently do plotting), we aren't returning anything, so I'm fine with changing this to whatever we need (batch containing images, ground truth, and predictions).

**Comment by DimitrisMantas** [comment]
> Maybe a base class for most shared code and a subclass for each logger?

I can add TensorBoard and MLFlow, as the default and (nice) alternative.

> What do the Lightning docs suggest for callbacks? Should they have access to arbitrary instance attributes, or is it better to return all required values?

I don't believe there is a consensus or general recommendation. However, there are definitely callbacks that assume the existence of certain instance attributes, such as `BackboneFinetuning` and `LRFinder`.

The callbacks have `on_<stage>_<batch/epoch>_end` methods that provide access to the associated to the batch `<stage>_step` receives as well as the whatever it returns. So, we either append whatever we need to to the batch during the step and keep returning `None`, or add explicit return values.

For metric plotting, I think it's best to expect specific attributes provided by the base task or the metric mixin.

For predictions, I think it's best to make samples/batches "immutable", return logits/predictions/etc. and have the callback construct a new sample and pass it to the data module plot method and finally log the resulting figure.

I'm thinking something like this:

```python
class Sample(TypedDict):
    image: ReadOnly[torch.Tensor]
    mask: ReadOnly[torch.Tensor]
    preds: NotRequired[ReadOnly[torch.Tensor]]

type Batch = Sample

class BasePlotter(lightning.pytorch.callbacks.Callback):
    def plot_sample(
        self,
        trainer: lightning.Trainer,
        pl_module: lightning.LightningModule,
        outputs: torch.Tensor | Mapping[str, Any] | None,
        batch: Batch,
        batch_idx: int,
        dataloader_idx: int = 0,
    ) -> plt.Figure | None:
        ...
```

In general, the callback would have to check whether logging is enabled, search through the associated loggers for the one it expects, and then plot whatever it needs to, but the specifics depend on what exactly it is we're plotting. The above snippet is for plotting predictions; I assume that plotting metrics would be easier...

**Comment by adamjstewart** [comment]
> So, we either append whatever we need to to the batch during the step and keep returning None, or add explicit return values.

I would prefer to add explicit return values. [Explicit is better than implicit](https://peps.python.org/pep-0020/).

> For metric plotting, I think it's best to expect specific attributes provided by the base task or the metric mixin.

How feasible would it be to compute a confusion matrix in the callback? Would each trainer be different enough that we would instead have to have trainer-specific logic? I'm also okay with your suggestion too.

> For predictions, I think it's best to make samples/batches "immutable", return logits/predictions/etc. and have the callback construct a new sample and pass it to the data module plot method and finally log the resulting figure.

Why would they need to be immutable? Currently we essentially have `type Sample = dict[str, Tensor]` and add predictions to the sample/batch before plotting. My only suggested change is to also return this batch.

**Comment by DimitrisMantas** [comment]
> How feasible would it be to compute a confusion matrix in the callback?

The base callback provides hooks at the start and end of each batch and epoch, so I assume it is possible, but I think the best option is to have the metric mixin handle it.

> Why would they need to be immutable?

They don't, but perhaps it's a bit confusing to have the sample/batch state change throughout execution. That said, there's only one change happening right now at a clearly defined point, so maybe it's fine.

> My only suggested change is to also return this batch.

We don't even need to return it, the callback gets a reference to it since its a non-primitive object, so it can also see any changes we make to it.

**Comment by adamjstewart** [comment]
We don't _need_ to, but explicit is better than implicit.

**Comment by adamjstewart** [comment]
Just realized this is a duplicate of #182, let's continue the discussion there.
